### PR TITLE
[pulseaudio] Allow flexible parameters to find a given pulseaudio device

### LINKED ISSUE #12555: [pulseaudio] Bug/Feature Request: Avoid random assignment of pulseaudio sinks to openhab things
Dear Maintainer of the PulseAudio binding,

Thank you very much for providing this binding.  When working with this binding, I encountered a random assignment of PulsAudio sinks to a particular OpenHAB thing under OpenHAB 3.2.0.  Probably this effect is at the border between bug report and feature request.

The present version of the PulseAudio binding supports an assignment of PulseAudio sinks to a particular Thing in OpenHAB based on the name of the PulseAudio sink:

`Thing sink          multiroom       "Snapcast"           @ "Room"       [name="alsa_card.pci-0000_00_1f.3", activateSimpleProtocolSink="true", simpleProtocolSinkPort=4711]`

While this kind of assignment works fine for PulseAudio sinks realized in hardware, it is not possible to uniquely assign a PulseAudio sink that was discovered via the PulseAudio module module-raop-discovery.

The reason is that PulseAudio fails to assign always the same name on such PulseAudio sinks. This random assignment occurs in two cases:

1. In case you have an ipv4 and ipv6 network enabled, the same PulseAudio sink is discovered twice. As the propagated name of the respective audio sink is identical, PulseAudio appends an enumerator value to the name to make the name unique. But it is only the name that is unique, not the assignment. For example, audio sink raop_output.room.local is detected via the ipv4 network. Then the ipv6 network causes creation of raop_output.room.local.2. This would not be so bad if it is ensured that ipv6 always receives the .2 postfix. Unfortunately, it cannot be ensured that the assignment order is always the same. Therefore, when creating a thing definition in OpenHAB, it is a random assignment whether the ipv4 route will be taken or the ipv6 route. One might argue that this is irrelevant as long as it works. However, in terms of defined network conditions it appears not to be recommandable that the administrator has no way to enforce OpenHAB to use exactly the PulseAudio sink as desired. It is in fact a random choice whether ipv4 or ipv6 is used when addressing the respective OpenHAB thing.

2. In case you have a RAOP playback device (e.g. an AirPlay Receiver) capable of playing back multiple streams to support speakers in multiple rooms the situation gets even worse. The RAOP playback device shares the same name for all of the provided AirPlay sinks. As reported above for ipv4 / ipv6, each separate AirPlay sink will receive an enumerater value appended to the same name. As the sequence in the discovery of these AirPlay sinks differs upon their detection, we have a random assignment in the OpenHAB things as we can only use the name of the audio sink for the assignment to an OpenHAB thing and the name may differ in the next discovery. Therefore, it is impossible to specify the room where audio is played. It is random.

In view of these issues, I recommend to expand the possibilities to assign a PulseAudio sink to an OpenHAB Thing such that a unique assignment is supported without causing a random selection of a PulseAudio sink will be assigned to an OpenHAB thing in the next discovery.

Let me suggest a possible solution:

a) The assignment between PulseAudio sink and Thing shall consider an optional parameter sinkNetworkType that needs to match the network type (ipv4 or ipv6) of the PulseAudio sink.

b) The assignment between PulseAudio sink and Thing shall consider an optional parameter sinkDescription that needs to match the description of the PulseAudio sink.

c) The assignment between PulseAudio sink and Thing shall consider that the name may only specify the base name of the respective PulseAudio sink without any enumerator. So when specifying the name "raop_output.room.local" it also matches "raop_output.room.local.<enumerator>". The parameters specificed in a) and b) will make the choice unique.

When considering these parameters in the OpenHAB thing definition of a PulseAudio sink, an assignment is possible that remains unique also in the next discovery.

Thank you very much for considering this Feature Request.

Best regards,
Peter

### PR DESCRIPTION
**Pull request by dalgwen**
Following the request of a user, who had difficulties with its pulseaudio devices and the random naming occuring (#12555) , I would like to add some (non breaking) possibilities in the way to choose a pulseaudio device corresponding to a thing description.

To identify the device on the pulseaudio server, with this PR, you can now use the description instead of the technical id (a.k.a. "name").
To filter furthermore, you can also use the parameter additionalFilters (optional regular expressions that need to match a property value of a device on the pulseaudio server).
As you can see in the related discussion, these two modifications were necessary to match an interesting use case (playing sound to airplay device with the help of the RAOP pulseaudio module as an intermediary)

Closes #12555

[A snapshot is here](https://github.com/dalgwen/openhab-addons/releases/tag/3.3.0_SNAPSHOT-pulseaudio_12555)

Signed-off-by: Gwendal Roulleau <gwendal.roulleau@gmail.com>

### CODE REVIEW

File: bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/PulseaudioClient.java [hunk 1]
```diff
@@ -258,15 +261,23 @@ public void sendCommand(String command) {
...
-            }
+    public @Nullable AbstractAudioDeviceConfig getGenericAudioItem(DeviceIdentifier deviceIdentifier) {
+        List<AbstractAudioDeviceConfig> matchingDevices = items.stream()
+                .filter(device -> device.matches(deviceIdentifier))
+                .sorted(Comparator.comparing(AbstractAudioDeviceConfig::getPaName)).collect(Collectors.toList());
+        if (matchingDevices.size() == 1) {
+            return matchingDevices.get(0);
+        } else if (matchingDevices.size() > 1) {
+            logger.debug(
+                    "Cannot select exactly one audio device, so, choosing the first. To choose without ambiguity between the {} devices matching the identifier {}, you can maybe use a more restrictive 'additionalFilter' parameter",
```
Comment (lolodomo) [inline bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/PulseaudioClient.java:278]: That is minor but I think there is a comma to suppress : ... audio device, so choosing the first.

**Review by lolodomo — CHANGES_REQUESTED** [review]
Just 2 minor comments

File: bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/handler/DeviceIdentifier.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+ * @author Gwendal Roulleau - Initial contribution
+ *
+ */
+@NonNullByDefault
+public class DeviceIdentifier {
+
+    private String nameOrDescription;
+    private List<Pattern> additionalFilters = new ArrayList<>();
+
+    public DeviceIdentifier(String nameOrDescription, @Nullable String additionalFilters) {
```
[the remark this answers is not in the document]
Comment (dalgwen) [reply bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/handler/DeviceIdentifier.java:36]: Indeed

**Comment by dalgwen** [comment]
Thank you again @lolodomo

**Review by lolodomo — APPROVED** [review]
LGTM, thank you

**Pull request merged**